File: 0040-combination-sum-ii/0040-combination-sum-ii.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum2 = function(candidates, target) {
    candidates.sort((a, b) => a - b);
    const result = [];
    backtrack(candidates, target, 0, [], result);
    return result;
};

function backtrack(candidates, target, start, path, result) {
    if (target === 0) {
        result.push([...path]);
        return;
    }

    for (let i = start; i < candidates.length; i++) {
        if (candidates[i] > target) break;

        // Skip the same number after the first one to avoid duplicates
        if (i > start && candidates[i] === candidates[i - 1]) continue;

        path.push(candidates[i]);
        backtrack(candidates, target - candidates[i], i + 1, path, result);
        path.pop();
    }
}

File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:
    def combinationSum2(self, candidates, target):
        candidates.sort()
        result = []
        self._combine(candidates, 0, [], target, result)
        return result

    def _combine(self, nums, start, path, target, result):
        if target == 0:
            result.append(path)
            return 
        if target < 0:
            return
        for i in range(start, len(nums)):
            if i > start and nums[i] == nums[i-1]:
                continue
            self._combine(nums, i+1, path + [nums[i]], target-nums[i], result)
